open_lm/file_utils.py

```diff
@@ -175,7 +181,76 @@ def source_exhausted(paths, shard_list_per_source):
     return False
 
 
-def get_string_for_epoch(num_samples, starting_chunk, paths, weights, min_shards_needed):
+def adjust_samples(shard_list, manifest, starting_index, num_workers, world_size):
+    """Adjust the number of samples to train on.
+
+    This function returns the number of samples that we will train on, so that no worker will keep going
+    while the others have stopped. This is calculated via the same assignment of nodes that wds.split_by_node
+    and wds.split_by_worker does.
```

Comment:
Good point. I'll rewrite this function to use `wds.split_by_node` and `wds.split_by_worker directly.

Reply:
I'll remove `adjust_samples`, since there is no point in keeping only the samples from the smallest worker - it's best to leave them to end their data.

Reply:
I'll update this to just count the number of samples in these shards, and then let the dataloader workers run out of data. That seems to be the cleanest solution.

Reply:
Since we no longer need to count the minimum number of shards, I just removed adjust_samples and incorporated its functionality elsewhere.